File: tests/csv-export.test.mjs
// buildCSV() coverage — the flat CSV export (P2-12).
//
// buildCSV(state) is a pure helper (the downloadCSV side that touches the DOM
// is not exercised here). We assert the header, one row per record across the
// three data types, and RFC-4180 escaping of awkward values.
import { describe, it, expect, beforeEach } from "vitest";
import { loadStore } from "./load-store.mjs";

let S;
beforeEach(() => { S = loadStore(); });

function rows(csv) { return csv.split("\r\n"); }

describe("buildCSV — structure", () => {
  it("emits a header and a row per intention / block / habit mark", () => {
    const state = {
      today: { dayKey: "2026-02-02", intentions: [{ id: "a", text: "ship", done: true }, { id: "b", text: "rest", done: false }], reflection: "" },
      days: { "2026-02-01": { intentions: [{ id: "c", text: "plan", done: true }], reflection: "" } },
      blocks: [{
        id: "blk", title: "Deep work", project: "Pauta", status: "done", reflection: "good",
        createdAt: S.tsFromDayKey("2026-02-02"),
        sessions: [{ startedAt: S.tsFromDayKey("2026-02-02"), endedAt: S.tsFromDayKey("2026-02-02") + 25 * 60000 }],
      }],
      habits: [{ id: "h", name: "água", log: { "2026-02-01": 3 }, respiros: { "2026-02-02": true } }],
    };
    const r = rows(S.buildCSV(state));
    expect(r[0]).toBe("type,date,title,project,minutes,count,status,note");
    expect(r).toContain("intention,2026-02-02,ship,,,,done,");
    expect(r).toContain("intention,2026-02-02,rest,,,,open,");
    expect(r).toContain("intention,2026-02-01,plan,,,,done,"); // archived day
    expect(r).toContain("block,2026-02-02,Deep work,Pauta,25,,done,good");
    expect(r).toContain("habit,2026-02-01,água,,,3,done,"); // count carried
    expect(r).toContain("habit,2026-02-02,água,,,,respiro,");
  });

  it("handles empty state — just the header row", () => {
    const csv = S.buildCSV({ today: { dayKey: "2026-01-01", intentions: [], reflection: "" }, days: {}, blocks: [], habits: [] });
    expect(rows(csv).length).toBe(1);
  });
});

describe("csvCell — escaping", () => {
  it("quotes and doubles quotes for commas, quotes and newlines", () => {
    expect(S.csvCell("plain")).toBe("plain");
    expect(S.csvCell("a,b")).toBe('"a,b"');
    expect(S.csvCell('say "hi"')).toBe('"say ""hi"""');
    expect(S.csvCell("line1\nline2")).toBe('"line1\nline2"');
    expect(S.csvCell(null)).toBe("");
  });

  it("escapes an intention whose text contains a comma", () => {
    const state = {
      today: { dayKey: "2026-03-03", intentions: [{ id: "x", text: "buy milk, eggs", done: false }], reflection: "" },
      days: {}, blocks: [], habits: [],
    };
    expect(S.buildCSV(state)).toContain('intention,2026-03-03,"buy milk, eggs",,,,open,');
  });
});

File: android-native/app/src/test/kotlin/com/pauta/app/service/CsvExportTest.kt
package com.pauta.app.service

import com.pauta.app.data.entity.FocusBlockEntity
import com.pauta.app.data.entity.FocusSessionEntity
import com.pauta.app.data.entity.HabitEntity
import com.pauta.app.data.entity.HabitLogEntity
import com.pauta.app.data.entity.HabitRespiroEntity
import com.pauta.app.data.entity.IntentionEntity
import com.pauta.app.domain.DateUtils
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** Tests the flat CSV export (ported from buildCSV in store.jsx). */
class CsvExportTest {

    @Test
    fun buildsHeaderRowsAndEscaping() {
        val created = DateUtils.msFromKey("2025-05-10")
        val data = BackupData(
            intentions = listOf(
                IntentionEntity(id = "i1", dayKey = "2025-05-10", text = "Estudar", done = true, createdAt = created),
                IntentionEntity(id = "i2", dayKey = "2025-05-10", text = "Ler, com vírgula", done = false, createdAt = created + 1),
            ),
            blocks = listOf(
                FocusBlockEntity(id = "b1", title = "Foco", project = "Escola", status = "done", reflection = "bom", createdAt = created),
            ),
            sessions = listOf(
                FocusSessionEntity(id = "s1", blockId = "b1", startedAt = created, endedAt = created + 25 * 60_000L),
            ),
            habits = listOf(HabitEntity(id = "h1", name = "Água")),
            logs = listOf(HabitLogEntity(habitId = "h1", dayKey = "2025-05-10")),
            respiros = listOf(HabitRespiroEntity(habitId = "h1", dayKey = "2025-05-11")),
        )

        val lines = CsvExport.build(data).split("\r\n")

        assertEquals("type,date,title,project,minutes,count,status,note", lines[0])
        assertTrue(lines.contains("intention,2025-05-10,Estudar,,,,done,"))
        // a comma in the text forces quoting
        assertTrue(lines.contains("intention,2025-05-10,\"Ler, com vírgula\",,,,open,"))
        // 25-minute session → 25 minutes
        assertTrue(lines.contains("block,2025-05-10,Foco,Escola,25,,done,bom"))
        assertTrue(lines.contains("habit,2025-05-10,Água,,,1,done,"))
        assertTrue(lines.contains("habit,2025-05-11,Água,,,,respiro,"))
    }
}
